File: win/mobiledevice/AMDevice.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using CoreFoundation;
using ExtensionMethods;
using PlistCS;

namespace mobiledevice
{

    class AMDevice
    {
        IntPtr am_device;
        string deviceId;
        int socket_fd = 0;

        public AMDevice(IntPtr device)
        {
            am_device = device;
            deviceId = MobileDevice.GetDeviceIdentifier(device);
        }

        public void WriteLine(string msg)
        {
            Console.Out.WriteLine(deviceId + " " + msg);
            Console.Out.Flush();
        }

        public bool Connect()
        {
            return MobileDevice.Connect(am_device);
        }
        public bool Disconnect()
        {
            return MobileDevice.Disconnect(am_device);
        }

        public bool ValidatePairing()
        {
            if ( MobileDevice.IsPaired(am_device) )
            {
                return MobileDevice.ValidatePairing(am_device);
            }
            else
            {
                return MobileDevice.PairDevice(am_device);
            }
        }

        public bool StartSession()
        {
            return MobileDevice.StartSession(am_device);
        }
        public bool StopSession()
        {
            return MobileDevice.StopSession(am_device);
        }

        public const string SVC_INSTALLATION_PROXY = "com.apple.mobile.installation_proxy";
        public const string SVC_MCINSTALL = "com.apple.mobile.MCInstall";
        public const string SVC_DIAGNOSTICS_RELAY = "com.apple.mobile.diagnostics_relay";
        public const string SVC_MISAGENT = "com.apple.misagent";

        public bool StartService(string svc)
        {
            CFString c = new CFString(svc);
            return MobileDevice.StartService(am_device, c.typeRef, out socket_fd);
        }

        public Hashtable LookupApps()
        {
            bool direct = true;
            if ( direct )
            {
                return MobileDevice.LookupApplication(am_device);
            }

            if ( !StartService(SVC_INSTALLATION_PROXY) )
            {
                return null;
            }

            Dictionary<string, object> send = new Dictionary<string, object>
            {
                { "Command","Lookup" }
            };
            Dictionary<string, object> received = TransferPlist(send);
            Flush();

            Dictionary<string, object> lookup = received["LookupResult"] as Dictionary<string, object>;
            Dictionary<string, object>.KeyCollection keys = lookup.Keys;

            string CFBundleIdentifier = MobileDevice.CFBundleIdentifier;
            string ApplicationType = MobileDevice.ApplicationType;
            string CFBundleName = MobileDevice.CFBundleName;
            string CFBundleDisplayName = MobileDevice.CFBundleDisplayName;
            string CFBundleShortVersionString = MobileDevice.CFBundleShortVersionString;
            string CFBundleVersion = MobileDevice.CFBundleVersion;

            Hashtable apps = new Hashtable();
            foreach ( string key in keys )
            {
                Dictionary<string, object> v = lookup[key] as Dictionary<string, object>;
                string bundle = v.Find(CFBundleIdentifier) as string;
                string type = v.Find(ApplicationType) as string;
                string name = v.Find(CFBundleName) as string;
                string displayname = v.Find(CFBundleDisplayName) as string;
                string version = v.Find(CFBundleShortVersionString) as string;
                string build = v.Find(CFBundleVersion) as string;
                Hashtable app = new Hashtable();
                app.Add(CFBundleIdentifier, bundle);
                app.Add(ApplicationType, type);
                app.Add(CFBundleName, name);
                app.Add(CFBundleDisplayName, displayname);
                app.Add(CFBundleShortVersionString, version);
                app.Add(CFBundleVersion, build);
                apps.Add(bundle, app);
            }
            return apps;
        }

        public void showApps(Hashtable apps)
        {
            string CFBundleIdentifier = MobileDevice.CFBundleIdentifier;
            string ApplicationType = MobileDevice.ApplicationType;
            string CFBundleName = MobileDevice.CFBundleName;
            string CFBundleDisplayName = MobileDevice.CFBundleDisplayName;
            string CFBundleShortVersionString = MobileDevice.CFBundleShortVersionString;
            string CFBundleVersion = MobileDevice.CFBundleVersion;

            foreach ( DictionaryEntry e in apps )
            {
                Hashtable app = e.Value as Hashtable;
                Console.Out.Write(app[CFBundleIdentifier]);
                Console.Out.Write("\t");
                Console.Out.Write(app[ApplicationType]);
                Console.Out.Write("\t");
                Console.Out.Write(app[CFBundleName]);
                Console.Out.Write(" ");
                Console.Out.Write(app[CFBundleShortVersionString]);
                Console.Out.Write("(");
                Console.Out.Write(app[CFBundleVersion]);
                Console.Out.Write(")");
                Console.Out.Write("\n");
            }
            Console.Out.Flush();
        }

        public bool InstallApp(string appPath)
        {
            FileInfo fd = new FileInfo(appPath);
            if ( !fd.Exists )
            {
                return false;
            }

            return MobileDevice.InstallApp(am_device, appPath);
        }

        public bool UninstallApp(string bundleId)
        {
            return MobileDevice.UninstallApp(am_device, bundleId);
        }

        Dictionary<string, object> TransferPlist(Dictionary<string, object> dict)
        {
            Dictionary<string, object> input = dict;
            byte[] source = Plist.writeBinary(input);
            byte[] dest = MobileDevice.SocketIO(socket_fd, source);
            Dictionary<string, object> output = Plist.readPlist(dest) as Dictionary<string, object>;
            return output;
        }

        public bool Flush()
        {
            Dictionary<string, object> dict = new Dictionary<string, object>
            {
                { "RequestType","Flush" }
            };
            Dictionary<string, object> received = TransferPlist(dict);
            return true;
        }

        public Hashtable LookupProfiles()
        {
            if ( !StartService(SVC_MCINSTALL) )
            {
                return null;
            }

            Dictionary<string, object> dict = new Dictionary<string, object>
            {
                { "RequestType","GetProfileList" }
            };
            Dictionary<string, object> received = TransferPlist(dict);
            Flush();

            string ProfileMetadata = MobileDevice.ProfileMetadata;
            string OrderedIdentifiers = MobileDevice.OrderedIdentifiers;
            string PayloadDisplayName = MobileDevice.PayloadDisplayName;
            string PayloadOrganization = MobileDevice.PayloadOrganization;
            string PayloadIdentifier = MobileDevice.PayloadIdentifier;

            List<object> keys = received[OrderedIdentifiers] as List<object>;
            int length = keys.Count;
            Dictionary<string, object> metas = received[ProfileMetadata] as Dictionary<string, object>;

            Hashtable profiles = new Hashtable();
            for ( int i = 0; i < length; i++ )
            {
                string identifier = keys[i] as string;
                Dictionary<string, object> meta = metas[identifier] as Dictionary<string, object>;
                string displayname = meta.Find(PayloadDisplayName) as string;
                string organization = meta.Find(PayloadOrganization) as string;
                Hashtable m = new Hashtable();
                m.Add(PayloadIdentifier, identifier);
                m.Add(PayloadDisplayName, displayname);
                m.Add(PayloadOrganization, organization);
                profiles.Add(identifier, m);
            }
            return profiles;
        }

        public void showProfiles(Hashtable profiles)
        {
            string ProfileMetadata = MobileDevice.ProfileMetadata;
            string OrderedIdentifiers = MobileDevice.OrderedIdentifiers;
            string PayloadDisplayName = MobileDevice.PayloadDisplayName;
            string PayloadOrganization = MobileDevice.PayloadOrganization;
            string PayloadIdentifier = MobileDevice.PayloadIdentifier;

            foreach ( DictionaryEntry e in profiles )
            {
                Hashtable app = e.Value as Hashtable;
                Console.Out.Write(app[PayloadIdentifier]);
                Console.Out.Write("\t");
                Console.Out.Write(app[PayloadDisplayName]);
                Console.Out.Write("(");
                Console.Out.Write(app[PayloadOrganization]);
                Console.Out.Write(")");
                Console.Out.Write("\n");
            }
            Console.Out.Flush();
        }

        public bool InstallProfile(string profilePath)
        {
            FileInfo fd = new FileInfo(profilePath);
            if ( !fd.Exists )
            {
                return false;
            }

            if ( !StartService(SVC_MCINSTALL) )
            {
                return false;
            }

            byte[] bytes = File.ReadAllBytes(profilePath);
            Dictionary<string, object> dict = new Dictionary<string, object>
            {
                { "RequestType","InstallProfile" },
                { "Payload",bytes },
            };

            Dictionary<string, object> received = TransferPlist(dict);
            string status = received["Status"] as string;
            Flush();

            return status.Equals("Acknowledged");
        }

        public bool UninstallProfile(string profileId)
        {
            if ( !StartService(SVC_MCINSTALL) )
            {
                return false;
            }

            Dictionary<string, object> dict = new Dictionary<string, object>
            {
                { "RequestType","RemoveProfile" },
                { "ProfileIdentifier",profileId },
            };
            Dictionary<string, object> received = TransferPlist(dict);
            string status = received["Status"] as string;
            Flush();

            return status.Equals("Acknowledged");
        }

        public bool Shutdown()
        {
            if ( !StartService(SVC_DIAGNOSTICS_RELAY) )
            {
                return false;
            }

            Dictionary<string, object> dict = new Dictionary<string, object>
            {
                { "Request","Shutdown" }
            };
            Dictionary<string, object> received = TransferPlist(dict);
            string status = received["Status"] as string;
            Flush();

            return status.Equals("Acknowledged");
        }

        public bool UpdateTime()
        {
            //DateTime dt = DateTime.Now;
            //DateTime ep = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
            //TimeSpan sp = dt.Subtract(ep);
            //double tk = sp.TotalSeconds;
            //return MobileDevice.updateTime(am_device, tk);

            //"com.apple.misagent",
            //"com.apple.mobile.diagnostics_relay",
            //"com.apple.mobile.MCInstall",
            //"com.apple.iosdiagnostics.relay",

            if ( !StartService(SVC_MISAGENT) )
            {
                return false;
            }

            DateTime dt = DateTime.Now;
            DateTime ep = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
            TimeSpan sp = dt.Subtract(ep);
            double tk = sp.TotalSeconds;
            Dictionary<string, object> dict = new Dictionary<string, object>
            {
                { "Request","SetValue" },
                { "Domain","NULL" },
                { "Key","TimeIntervalSince1970" },
                { "Value",tk }
            };
            Dictionary<string, object> received = TransferPlist(dict);
            string status = received["Status"].ToString();
            //Flush();

            return status.Equals("0");
        }

    }

}

File: osx/mobiledevice/AMDevice.m
//
//  AMDevice.m
//  mobiledevice
//
//  Created by huangyi on 15/8/31.
//  Copyright (c) 2015年 wettags. All rights reserved.
//

#import "AMDevice.h"

#import "CoreFoundation.h"
#import "MobileDevice.h"



@interface AMDevice ()

@end

@implementation AMDevice

+(AMDevice *)deviceWithHandle:(struct am_device *)device{
    AMDevice *dev=[[AMDevice alloc] init];
    dev.device=device;
    return dev;
}

-(void)setDevice:(struct am_device *)device{
    _device=device;
    CFStringRef identifier=AMDeviceCopyDeviceIdentifier(device);
    NSString *deviceId=[NSString stringWithString:(__bridge NSString *)(identifier)];
    _device_id=deviceId;
}


-(NSString*)deviceId{
    return self.device_id;
}

-(void)WriteLine:(NSString *)message{
    NSString *msg=[NSString stringWithFormat:@"%@ %@",self.device_id,message];
    WriteLine(msg);
}


-(BOOL)Connect{
    return Connect(self);
}
-(BOOL)Disconnect{
    return Disconnect(self);
}
-(BOOL)ValidatePairing{
    return ValidatePairing(self);
}
-(BOOL)StartSession{
    return StartSession(self);
}
-(BOOL)StopSession{
    return StopSession(self);
}

-(BOOL)StartService:(const NSString *)serviceName{
    return StartService(self, serviceName);
}
-(NSDictionary*)TransferPlist:(NSDictionary*)dict{
    if (!dict) {
        return nil;
    }
    NSError *error = nil;
    NSPropertyListFormat format = NSPropertyListXMLFormat_v1_0;
    NSData *source = [NSPropertyListSerialization dataWithPropertyList:dict format:format options:NSPropertyListWriteStreamError error:&error];
    NSData *dest=SocketIO(self, source);
    NSDictionary *output = [NSPropertyListSerialization propertyListWithData:dest options:NSPropertyListReadStreamError format:&format error:&error];
    return output;
}
-(BOOL)Flush{
    NSDictionary *dict=@{ @"RequestType":@"Flush" };
    NSDictionary *received=[self TransferPlist:dict];
    if (!received) {
        return false;
    }
    return true;
}


-(NSDictionary *)LookupApps{
    return LookupApps(self);
}

-(BOOL)InstallApp:(NSString *)appPath{
    if (!IsFileExists(appPath)) {
        return false;
    }
    return InstallApp(self, appPath);
}
-(BOOL)UninstallApp:(NSString *)appId{
    return UninstallApp(self, appId);
}


-(NSDictionary *)LookupProfiles{
    if (![self StartService:SVC_MCINSTALL]) {
        return nil;
    }
    NSDictionary *dict=@{ @"RequestType":@"GetProfileList" };
    NSDictionary *profiles=[self TransferPlist:dict];
    NSDictionary *metas=profiles[ProfileMetadata];
    return metas;
}

-(BOOL)InstallProfile:(NSString *)profilePath{
    if (!IsFileExists(profilePath)) {
        return false;
    }
    if (![self StartService:SVC_MCINSTALL]) {
        return false;
    }
    NSData *data=[NSData dataWithContentsOfFile:profilePath];
    NSDictionary *dict=@{ @"RequestType":@"InstallProfile",@"Payload":data };
    NSDictionary *received=[self TransferPlist:dict];
    NSString *status=received[@"Status"];
    [self Flush];
    return [status isEqual:@"Acknowledged"];
}
-(BOOL)UninstallProfile:(NSString *)profileId{
    if (![self StartService:SVC_MCINSTALL]) {
        return false;
    }
    NSDictionary *dict=@{ @"RequestType":@"RemoveProfile",@"ProfileIdentifier":profileId };
    NSDictionary *received=[self TransferPlist:dict];
    NSString *status=received[@"Status"];
    [self Flush];
    return [status isEqual:@"Acknowledged"];
}


-(BOOL)Shutdown{
    if (![self StartService:SVC_DIAGNOSTICS_RELAY]) {
        return false;
    }
    NSDictionary *dict=@{ @"Request":@"Shutdown" };
    NSDictionary *received=[self TransferPlist:dict];
    NSString *status=received[@"Status"];
    [self Flush];
    return [status isEqual:@"Acknowledged"];
}
-(BOOL)UpdateTime{
    if (![self StartService:SVC_MISAGENT]) {
        return false;
    }
    NSTimeInterval tk=[[NSDate date] timeIntervalSince1970];
    NSDictionary *dict=@{ @"Request":@"SetValue", @"Domain":@(0), @"Key":@"TimeIntervalSince1970", @"Value":@(tk) };
    NSDictionary *received=[self TransferPlist:dict];
    NSString *status=[NSString stringWithFormat:@"%@",received[@"Status"]];
    [self Flush];
    return [status isEqual:@"0"];
}

@end
